From 41188b89d35a4c8b462de01f53fbd5f382aed6bc Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 12:01:27 -0400
Subject: [ethernet] mlx4: IRQ affinity hint is not cleared on port down

Message-id: <67edf5149623350afb62dffd6005314c46907bb4.1410425016.git.avadai@redhat.com>
Patchwork-id: 93704
O-Subject: [Patch RHEL7.1 V1 151/197] net/mlx4_en: IRQ affinity hint is not cleared on port down
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Amir Vadai <amirv@mellanox.com>

Need to remove affinity hint at mlx4_en_deactivate_cq() and not at
mlx4_en_destroy_cq() - since affinity_mask might be free'd while still
being used by procfs.

Signed-off-by: Amir Vadai <amirv@mellanox.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit bb273617a65b9ed75f8cf9417206cbfcfb41fc48)
Signed-off-by: Amir Vadai <avadai@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/mellanox/mlx4/en_cq.c b/drivers/net/ethernet/mellanox/mlx4/en_cq.c
index 5a280cd..6b2e160 100644
--- a/drivers/net/ethernet/mellanox/mlx4/en_cq.c
+++ b/drivers/net/ethernet/mellanox/mlx4/en_cq.c
@@ -194,8 +194,6 @@ void mlx4_en_destroy_cq(struct mlx4_en_priv *priv, struct mlx4_en_cq **pcq)
  mlx4_en_unmap_buffer(&cq->wqres.buf);
  mlx4_free_hwq_res(mdev->dev, &cq->wqres, cq->buf_size);
  if (priv->mdev->dev->caps.comp_pool && cq->vector) {
-  if (!cq->is_tx)
-   irq_set_affinity_hint(cq->mcq.irq, NULL);
   mlx4_release_eq(priv->mdev->dev, cq->vector);
  }
  cq->vector = 0;
@@ -211,6 +209,7 @@ void mlx4_en_deactivate_cq(struct mlx4_en_priv *priv, struct mlx4_en_cq *cq)
  if (!cq->is_tx) {
   napi_hash_del(&cq->napi);
   synchronize_rcu();
+  irq_set_affinity_hint(cq->mcq.irq, NULL);
  }
  netif_napi_del(&cq->napi);
 
-- 
1.7.1